### ISSUE TITLE
[3.50.0-beta.9] Mesh

### ISSUE THREAD

**Issue by MadDogMayCry0** [photonstorm/phaser#5380]
https://codesandbox.io/s/clever-swanson-7hfw5
https://7hfw5.csb.app/
http://doctoral-arrangemen.000webhostapp.com/mesh/

The mesh is not displayed. I tried many different ways, but the result is always the same - the mesh does not appear on the stage.

The sample code is taken from here
https://www.patreon.com/posts/42666302
![image](https://user-images.githubusercontent.com/2493591/97786299-dc56d300-1baa-11eb-876f-111da8ad3510.png)

**Comment by photonstorm** [comment]
You have to use the latest build from master. Or Beta 10 will work fine, too. Here's a fully working example:

```js
class Example extends Phaser.Scene
{
    constructor ()
    {
        super();
    }

    preload ()
    {
        this.load.image('bg', 'assets/skies/gradient13.png');
        this.load.image('ayu', 'assets/pics/ayu.png');
    }

    create ()
    {
        this.add.image(400, 300, 'bg').setFlip(false, true);

        const vertices = [
            -1, 1,
            1, 1,
            -1, -1,
            1, -1
        ];

        const uvs = [
            0, 0,
            1, 0,
            0, 1,
            1, 1
        ];

        const indicies = [ 0, 2, 1, 2, 3, 1 ];

        const mesh = this.add.mesh(400, 300, 'ayu');

        mesh.addVertices(vertices, uvs, indicies);

        mesh.panZ(7);

        this.debug = this.add.graphics();

        mesh.setDebug(this.debug);

        this.add.text(16, 16, 'Rotate with mouse (+ Shift to pan)\nWheel to zoom\nD to toggle debug');

        this.input.keyboard.on('keydown-D', () => {

            if (mesh.debugCallback)
            {
                mesh.setDebug();
            }
            else
            {
                mesh.setDebug(this.debug);
            }

        });

        const rotateRate = 1;
        const panRate = 1;
        const zoomRate = 4;

        this.input.on('pointermove', pointer => {

            if (!pointer.isDown)
            {
                return;
            }

            if (!pointer.event.shiftKey)
            {
                mesh.modelRotation.y += pointer.velocity.x * (rotateRate / 800);
                mesh.modelRotation.x += pointer.velocity.y * (rotateRate / 600);
            }
            else
            {
                mesh.panX(pointer.velocity.x * (panRate / 800));
                mesh.panY(pointer.velocity.y * (panRate / 600));
            }

        });

        this.input.on('wheel', (pointer, over, deltaX, deltaY, deltaZ) => {

            mesh.panZ(deltaY * (zoomRate / 600));

        });
    }

    update ()
    {
        this.debug.clear();
        this.debug.lineStyle(1, 0x00ff00);
    }
}

const config = {
    type: Phaser.AUTO,
    width: 800,
    height: 600,
    backgroundColor: '#0a440a',
    parent: 'phaser-example',
    scene: Example
};

let game = new Phaser.Game(config);
```

You can find more here: http://labs.phaser.io/index.html?dir=3.50/game%20objects/mesh/&q=

**Comment by MadDogMayCry0** [comment]
```Javascript
var SceneMain = new Phaser.Scene('SceneMain');

SceneMain.preload = function(){   
	this.load.image('ball', 'ball.png');
}

SceneMain.create = function(){
	const vertices = [
		-1,1,
		1,1,
		-1,-1,
		1,-1];
    const uvs = [
		0,0,
		1,0,
		0,1,
		1,1]; 
    const indicies = [ 0, 2, 1, 2, 3, 1 ];
	const mesh = this.add.mesh(400, 300, "ball");
	mesh.addVertices(vertices, uvs, indicies);
}


const config = {
	type: Phaser.WEBGL,
	scale: {
		mode: Phaser.Scale.FIT,
		autoCenter: Phaser.Scale.CENTER_BOTH,
		width: 1920,
		height: 1080
	},
	audio: {
		noAudio: true
	},
	scene: [SceneMain]
};

var game = new Phaser.Game(config);
```
phaser.js:35462 Uncaught TypeError: Cannot read property 'length' of undefined at new Mesh (phaser.js:35462)
```Javascript
function Mesh (scene, x, y, vertices, uv, colors, alphas, texture, frame)
    {
        GameObject.call(this, scene, 'Mesh');

        if (vertices.length !== uv.length)
        {
            throw new Error('Mesh Vertex count must match UV count');
        }

```

**Comment by lackhand** [comment]
I ran into this also and am driveby documenting my findings here for @MadDogMayCry0 since there _is_ a logical explanation.

TL;DR: I think that this isn't really a bug; `mesh` objects require you to specify camera geometry per object, which can be surprising.

The `mesh` game object is often constructed from `x` & `y` vertices, embedded in 3-space @ `z=0` ([see code](https://github.com/photonstorm/phaser/blob/d67c93646cbfe8ba5997954f8af29a8f3f9e8e91/src/geom/mesh/GenerateVerts.js#L116)). However, this doesn't actually display by default without further method calls.

Like the code says, ["Note that rendering still takes place using Phasers orthographic camera. As a result, all depth and face tests are done in orthographic space."](https://github.com/photonstorm/phaser/blob/d67c93646cbfe8ba5997954f8af29a8f3f9e8e91/src/gameobjects/mesh/Mesh.js#L32). However, that's a little misleading, because the rendering is also passed through the `mesh`'s [projection matrix](https://github.com/photonstorm/phaser/blob/d67c93646cbfe8ba5997954f8af29a8f3f9e8e91/src/gameobjects/mesh/Mesh.js#L298) which is initialized to an [un-phaser-like perspective camera with undocumented, but reasonable, defaults](https://github.com/photonstorm/phaser/blob/d67c93646cbfe8ba5997954f8af29a8f3f9e8e91/src/gameobjects/mesh/Mesh.js#L353).

I've drafted https://github.com/photonstorm/phaser/pull/5498 to address my understanding/prevent others from getting confused. (it might be nice to go further and change defaults? That's probably too hard though). Feelsweird to do it on a closed bug, but I had the same confusion, so... :)

**Comment by MadDogMayCry0** [comment]
@lackhand thx for you answer.
